### PR TITLE
Create input-output interface in STM using Restartable

### LINKED ISSUE #24178: Bring Surrogate Model restart functionalities closer to the restartable functionalities in the framework
## Reason
The surrogate model restart capabilities use lower-level calls to the app's restartable functions. This separation might increase the maintenance burdens in the future.

## Design
Slightly extend the Restartable class and use it directly in the stochastic tools module.

## Impact
Better long-term maintainability

### PR DESCRIPTION
**Pull request by grmnptr**
Closes #24178

### CODE REVIEW

File: framework/include/restart/Restartable.h [hunk 1]
```diff
@@ -72,6 +73,19 @@ class Restartable
...
+   * Declare a piece of data as "restartable" and initialize it
+   * Similar to `declareRestartableData` but returns a const reference to the object.
+   *
+   * NOTE: This returns a _ const reference_!  Make sure you store it in a _const reference_!
+   *
+   * @param data_name The name of the data (usually just use the same name as the member variable)
+   * @param args Arguments to forward to the constructor of the data
+   */
+  template <typename T, typename... Args>
+  const T & getRestartableData(const std::string & data_name, Args &&... args) const;
```
Comment (lindsayad) [inline framework/include/restart/Restartable.h:86]: This is what I wanted for https://github.com/idaholab/moose/issues/21011 !

Comment (loganharbour) [inline framework/include/restart/Restartable.h:86]: Do we need to have constructor arguments for a getter?

I would think that if we have an object that is getting something but not declaring it, it must be trivial enough to have a default constructor. Even if this is in a restart where nothing ever declares it, and we're only getting

File: framework/include/restart/Restartable.h [hunk 2]
```diff
@@ -56,9 +56,10 @@ class Restartable
   Restartable(MooseApp & moose_app,
               const std::string & name,
               const std::string & system_name,
-              THREAD_ID tid);
+              THREAD_ID tid,
+              const bool read_only = false,
+              const RestartableDataMapName & metaname = "");
```
Comment (loganharbour) [inline framework/include/restart/Restartable.h:61]: We need to add detail in the doxygen here as it's getting pretty detailed

File: framework/include/restart/Restartable.h [hunk 3]
```diff
@@ -87,6 +101,21 @@ class Restartable
...
+   * Similar to `declareRestartableDataWithContext` but returns a const reference to the object.
+   *
+   * NOTE: This returns a _ const reference_!  Make sure you store it in a _const reference_!
+   *
+   * @param data_name The name of the data (usually just use the same name as the member variable)
+   * @param context Context pointer that will be passed to the load and store functions
+   * @param args Arguments to forward to the constructor of the data
+   */
+  template <typename T, typename... Args>
+  const T & getRestartableDataWithContext(const std::string & data_name,
```
Comment (loganharbour) [inline framework/include/restart/Restartable.h:115]: Does your use case need context @grmnptr? As this is a very advanced cased (where you need to get something from a whole different object, with the knowledge of how it was declared previously), I would prefer to keep the interface locked down for now and only allow this for getting this _without_ context until the need arrives.

File: framework/include/restart/Restartable.h [hunk 4]
```diff
@@ -177,11 +222,37 @@ Restartable::declareRestartableData(const std::string & data_name, Args &&... ar
...
+}
+
+template <typename T, typename... Args>
+const T &
+Restartable::getRestartableDataWithContext(const std::string & data_name,
+                                           void * context,
+                                           Args &&... args) const
+{
+  return std::as_const(
+      declareRestartableDataHelper<T>(data_name, context, std::forward<Args>(args)...).get());
```
Comment (loganharbour) [inline framework/include/restart/Restartable.h:248]: Look at https://stackoverflow.com/questions/123758/how-do-i-remove-code-duplication-between-similar-const-and-non-const-member-func

This is how this should be implemented. I think I led you poorly.

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: framework/include/restart/Restartable.h (hunk 3 above)
Comment (grmnptr) [reply framework/include/restart/Restartable.h:115]: Makes sense

File: framework/include/restart/Restartable.h (hunk 4 above)
Comment (grmnptr) [reply framework/include/restart/Restartable.h:248]: Sounds good to me!

**Comment by grmnptr** [comment]
What is you opinion @zachmprince ?

File: framework/include/restart/Restartable.h [hunk 5]
```diff
@@ -72,6 +82,19 @@ class Restartable
   template <typename T, typename... Args>
   T & declareRestartableData(const std::string & data_name, Args &&... args);
 
+  /**
+   * Declare a piece of data as "restartable" and initialize it
+   * Similar to `declareRestartableData` but returns a const reference to the object
+   * and does not allow forwarded arguments.
```
Comment (loganharbour) [inline framework/include/restart/Restartable.h:88]: Let's comment as to why forwarded arguments aren't needed

File: modules/stochastic_tools/include/interfaces/RestartableModelInterface.h [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  const T & getModelData(const std::string & data_name, Args &&... args) const;
+  ///@}
+
+  const std::string & modelMetaDataName() const { return _model_meta_data_name; }
+
+  const std::string _model_meta_data_name;
+
+private:
+  /// Reference to FEProblemBase instance
+  const MooseObject * _object;
```
Comment (loganharbour) [inline modules/stochastic_tools/include/interfaces/RestartableModelInterface.h:52]:
```suggestion
  /// Reference to the MooseObject that uses this interfface
  const MooseObject & _object;
```

File: modules/stochastic_tools/include/interfaces/RestartableModelInterface.h [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  /**
+   * Retrieve model data from the interface
+   */
+  template <typename T, typename... Args>
+  const T & getModelData(const std::string & data_name, Args &&... args) const;
+  ///@}
+
+  const std::string & modelMetaDataName() const { return _model_meta_data_name; }
+
+  const std::string _model_meta_data_name;
```
Comment (loganharbour) [inline modules/stochastic_tools/include/interfaces/RestartableModelInterface.h:48]: both need docstrings

File: modules/stochastic_tools/include/interfaces/RestartableModelInterface.h [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+private:
+  /// Reference to FEProblemBase instance
+  const MooseObject * _object;
+
+  Restartable _restartable;
```
Comment (loganharbour) [inline modules/stochastic_tools/include/interfaces/RestartableModelInterface.h:54]:
```suggestion
  Restartable _model_restartable;
```

**Review by loganharbour — CHANGES_REQUESTED** [review]
Just about there. I also have a solution to take care of the public getters that are a bit dirty.

File: framework/include/restart/Restartable.h [hunk 9]
```diff
@@ -156,18 +180,34 @@ class Restartable
   /// Flag for toggling read only status (see ReporterData)
   const bool _restartable_read_only;
 
+  /// Restartable metadata name
+  const RestartableDataMapName _metaname;
```
Comment (loganharbour) [inline framework/include/restart/Restartable.h:184]: This should be private

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: framework/include/restart/Restartable.h (hunk 9 above)
Comment (grmnptr) [reply framework/include/restart/Restartable.h:185]: Private MetaName reporting for duty!

**Review by loganharbour — APPROVED** [review]

**Pull request merged**